tensorflow/java/src/gen/cc/op_specs.h

```diff
@@ -0,0 +1,156 @@
+  EndpointSpec(const string& package, const string& name,
+      const Javadoc& javadoc, bool deprecated)
+    : package_(package), name_(name), javadoc_(javadoc),
+      deprecated_(deprecated) {}
+
+  const string& package() const { return package_; }
+  const string& name() const { return name_; }
+  const Javadoc& javadoc() const { return javadoc_; }
+  bool deprecated() const { return deprecated_; }
+
+ private:
+  const string package_;
+  const string name_;
+  const Javadoc javadoc_;
+  const bool deprecated_;
+};
+
+class ArgumentSpec {
+ public:
+  // A specification for an operation argument
+  //
+  // op_def_name: argument name, as known by TensorFlow core
+  // var: a variable to represent this argument in Java
+  // type: the tensor type of this argument
+  // description: a description of this argument, in javadoc
+  // iterable: true if this argument is a list
+  ArgumentSpec(const string& op_def_name, const Variable& var,
+      const Type& type, const string& description, bool iterable)
+    : op_def_name_(op_def_name), var_(var), type_(type),
+      description_(description), iterable_(iterable) {}
+  virtual ~ArgumentSpec() = default;
+
+  const string& op_def_name() const { return op_def_name_; }
+  const Variable& var() const { return var_; }
+  const Type& type() const { return type_; }
+  const string& description() const { return description_; }
+  bool iterable() const { return iterable_; }
+
+ private:
+  const string op_def_name_;
+  const Variable var_;
+  const Type type_;
+  const string description_;
+  const bool iterable_;
+};
+
+class AttributeSpec : public ArgumentSpec {
+ public:
+  // A specification for an operation attribute
+  //
+  // op_def_name: attribute name, as known by TensorFlow core
+  // var: a variable to represent this attribute in Java
+  // type: the type of this attribute
+  // jni_type: the type of this attribute in JNI layer (see OperationBuilder)
+  // description: a description of this attribute, in javadoc
+  // iterable: true if this attribute is a list
+  // has_default_value: true if this attribute has a default value if not set
+  AttributeSpec(const string& op_def_name, const Variable& var,
+      const Type& type, const Type& jni_type, const string& description,
+      bool iterable, bool has_default_value)
+    : ArgumentSpec(op_def_name, var, type, description, iterable),
+      jni_type_(jni_type), has_default_value_(has_default_value) {}
+  virtual ~AttributeSpec() = default;
+
+  const Type& jni_type() const { return jni_type_; }
+  bool has_default_value() const { return has_default_value_; }
+
+ private:
+  const Type jni_type_;
+  const bool has_default_value_;
+};
+
+class OpSpec {
+ public:
+  // Parses an op definition and its API to produce a specification used for
+  // rendering its Java wrapper
+  //
+  // op_def: Op definition
+  // api_def: Op API definition
+  static OpSpec Create(const OpDef& op_def, const ApiDef& api_def);
```

Comment:
There are a few reasons why it is better to do the ops generation in two phases (i.e. the parsing and the rendering). The main one is generics.

In order to know which attribute of an op is being used as a generic type for an argument, and which are gonna be inferred by inputs while others needs to be explicitly provided by the user, we need to go through all arguments/attributes of the ops before rendering the class.

Also, for each attributes identified to be generic, we need to keep track of its parameter name (e.g. the letter `U`) we assigned to it for other references, so there is a need to save parsed values in a stateful object like `*Spec`.

This is just an example, we also benefit from resolving only once arg/attr Java types and their Javadoc, since they are reused multiple times in the rendering phase.

Reply:
Hmm...I'm a little worried about the overall complexity here, but I admit I don't have a concrete suggestion on making this better. I guess keeping this `OpSpec` is fine for now.